kratos/controllers/controller.h

```diff
@@ -0,0 +1,175 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Daniel Diez
+//
+
+#pragma once
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "containers/model.h"
+#include "includes/kratos_flags.h"
+#include "includes/kratos_parameters.h"
+#include "includes/registry.h"
+
+namespace Kratos
+{
+
+///@name Kratos Classes
+///@{
+
+/**
+ * @class Controller
+ * @ingroup KratosCore
+ * @brief The base  class for all Controllers in Kratos.
+ * @details Controller is the base class for all controllers and defines the interface for them.
+ * Each class must define their own Evaluate method.
+ * Controller parameters must be passed at construction time.
+  @author Daniel Diez
+*/
+class Controller
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of Controller
+    KRATOS_CLASS_POINTER_DEFINITION(Controller);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    explicit Controller() = default;
+
+    /// Destructor.
+    virtual ~Controller() {}
+
+    /// Copy constructor.
+    //TODO: Check. It is required by the registry
+    Controller(Controller const& rOther) {}
+
+    /// Move constructor
+    Controller(Controller&& rOther) = default;
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    /// Assignment operator.
+    Controller& operator=(Controller const& rOther) = delete;
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief This method creates an pointer of the controlle
+     * @details We consider as input a Model and a set of Parameters for the sake of generality
+     * @warning Must be overrided in controller implementation
+     * @param rModel The model to be consider
+     * @param ThisParameters The configuration parameters
+     */
+    virtual Controller::Pointer Create(
+        Model& rModel,
+        Parameters ThisParameters) const
+    {
+        KRATOS_ERROR << "Calling base class Create. Please override this method in the corresonding Controller" << std::endl;
```

Comment:
why not make these functions pure virtual?

Reply:
Well, I think that if I make this class abstract, we cannot register it or export it to python. If you are ok with this I will make these functions pure virtual @matekelemen @philbucher @rubenzorrilla @sunethwarna

Reply:
> if I make this class abstract, we cannot register it or export it to python.

You'll need to define a trampoline class that handles the pure virtual functions:
- [pybind docs](https://pybind11.readthedocs.io/en/stable/advanced/classes.html)
- [example in kratos](https://github.com/KratosMultiphysics/Kratos/blob/b04aa01d2f368c51db283cd501aafe7e9b42384b/applications/HDF5Application/custom_python/add_custom_io_to_python.cpp#L54)

I think it's worth putting in the effort because this is supposed to be a widely used base class.

Reply:
Technically this is a PR

Reply:
works like a charm!

Reply:
thx!